File: core/src/com/bombbird/terminalcontrol/utilities/saving/FileLoader.java
package com.bombbird.terminalcontrol.utilities.saving;

import com.badlogic.gdx.Application;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.files.FileHandle;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.Base64Coder;
import com.bombbird.terminalcontrol.TerminalControl;
import com.bombbird.terminalcontrol.entities.airports.Airport;
import com.bombbird.terminalcontrol.entities.approaches.ILS;
import com.bombbird.terminalcontrol.entities.Runway;
import com.bombbird.terminalcontrol.entities.approaches.LDA;
import com.bombbird.terminalcontrol.entities.obstacles.Obstacle;
import com.bombbird.terminalcontrol.entities.procedures.HoldingPoints;
import com.bombbird.terminalcontrol.entities.procedures.MissedApproach;
import com.bombbird.terminalcontrol.entities.sidstar.Sid;
import com.bombbird.terminalcontrol.entities.sidstar.Star;
import com.bombbird.terminalcontrol.entities.waypoints.Waypoint;
import com.bombbird.terminalcontrol.entities.obstacles.PolygonObstacle;
import com.bombbird.terminalcontrol.entities.obstacles.CircleObstacle;
import com.bombbird.terminalcontrol.utilities.ErrorHandler;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;

public class FileLoader {
    public static String mainDir;

    public static Array<Obstacle> loadObstacles() {
        FileHandle obstacles = Gdx.files.internal("game/" + TerminalControl.radarScreen.mainName + "/" + TerminalControl.radarScreen.airac + "/obstacle.obs");
        Array<Obstacle> obsArray = new Array<>();
        String[] indivObs = obstacles.readString().split("\\r?\\n");
        for (String s: indivObs) {
            if ("".equals(s)) break;
            if (s.charAt(0) == '#') continue;
            //For each individual obstacle:
            PolygonObstacle obs = new PolygonObstacle(s);
            obsArray.add(obs);
            TerminalControl.radarScreen.stage.addActor(obs);
        }

        FileHandle restrictions = Gdx.files.internal("game/" + TerminalControl.radarScreen.mainName + "/" + TerminalControl.radarScreen.airac + "/restricted.restr");
        String[] indivRests = restrictions.readString().split("\\r?\\n");
        for (String s: indivRests) {
            if ("".equals(s)) break;
            if (s.charAt(0) == '#') continue;
            //For each individual restricted area
            CircleObstacle area = new CircleObstacle(s);
            obsArray.add(area);
            TerminalControl.radarScreen.stage.addActor(area);
        }
        return obsArray;
    }

    public static HashMap<String, Waypoint> loadWaypoints() {
        FileHandle handle = Gdx.files.internal("game/" + TerminalControl.radarScreen.mainName + "/" + TerminalControl.radarScreen.airac + "/waypoint.way");
        String wayptStr = handle.readString();
        String[] indivWpt = wayptStr.split("\\r?\\n");
        HashMap <String, Waypoint> waypoints = new HashMap<>(indivWpt.length + 1, 0.999f);
        for (String s: indivWpt) {
            //For each waypoint
            int index = 0;
            String name = "";
            int x = 0;
            int y = 0;
            for (String s1: s.split(" ")) {
                switch (index) {
                    case 0: name = s1; break;
                    case 1: x = Integer.parseInt(s1); break;
                    case 2: y = Integer.parseInt(s1); break;
                    default: Gdx.app.log("Load error", "Unexpected additional parameter in game/" + TerminalControl.radarScreen.mainName + "/" + TerminalControl.radarScreen.airac + "/restricted.rest");
                }
                index++;
            }
            Waypoint waypoint = new Waypoint(name, x, y);
            waypoints.put(name, waypoint);
            TerminalControl.radarScreen.stage.addActor(waypoint);
        }
        return waypoints;
    }

    public static HashMap<String, Runway> loadRunways(String icao) {
        HashMap<String, Runway> runways = new HashMap<>();
        FileHandle handle = Gdx.files.internal("game/" + TerminalControl.radarScreen.mainName + "/" + TerminalControl.radarScreen.airac + "/runway" + icao + ".rwy");
        String[] indivRwys = handle.readString().split("\\r?\\n");
        for (String s: indivRwys) {
            //For each individual runway
            Runway runway = new Runway(s);
            runways.put(runway.getName(), runway);
        }
        return runways;
    }

    public static HashMap<String, Star> loadStars(Airport airport) {
        //Load STARs
        HashMap<String, Star> stars = new HashMap<>();
        FileHandle handle = Gdx.files.internal("game/" + TerminalControl.radarScreen.mainName + "/" + TerminalControl.radarScreen.airac + "/star" + airport.getIcao() + ".star");
        JSONObject jo = new JSONObject(handle.readString());
        for (String name: jo.keySet()) {
            //For each individual STAR
            Star star = new Star(airport, jo.getJSONObject(name));
            star.setName(name);
            stars.put(name, star);
        }
        return stars;
    }

    public static HashMap<String, Sid> loadSids(Airport airport) {
        //Load SIDs
        HashMap<String, Sid> sids = new HashMap<>();
        FileHandle handle = Gdx.files.internal("game/" + TerminalControl.radarScreen.mainName + "/" + TerminalControl.radarScreen.airac + "/sid" + airport.getIcao() + ".sid");
        JSONObject jo = new JSONObject(handle.readString());
        for (String name: jo.keySet()) {
            //For each individual SID
            Sid sid = new Sid(airport, jo.getJSONObject(name));
            sid.setName(name);
            sids.put(name, sid);
        }
        return sids;
    }

    public static HashMap<String, int[]> loadAircraftData() {
        HashMap<String, int[]> aircrafts = new HashMap<>();
        FileHandle handle = Gdx.files.internal("game/aircrafts/aircrafts.air");
        String[] indivAircrafts = handle.readString().split("\\r?\\n");
        for (String s: indivAircrafts) {
            if (s.charAt(0) == '#') {
                continue;
            }
            //For each aircraft
            int index = 0;
            String icao = "";
            int[] perfData = new int[6];
            for (String s1: s.split(",")) {
                if (index == 0) {
                    icao = s1; //Icao code of aircraft
                } else if (index >= 2 && index <= 5) {
                    perfData[index - 1] = Integer.parseInt(s1);
                } else if (index == 1 || index == 6) {
                    //Wake turb cat, recat
                    char cat = s1.charAt(0);
                    perfData[index - 1] = cat;
                    if (index == 1 && !(cat == 'M' || cat == 'H' || cat == 'J')) Gdx.app.log("Wake cat error", "Unknown category for " + icao);
                    if (index == 6 && !(cat >= 'A' && cat <= 'F')) Gdx.app.log("Recat error", "Unknown category for " + icao);
                } else {
                    Gdx.app.log("Load error", "Unexpected additional parameter in game/aircrafts.air");
                }
                index++;
            }
            aircrafts.put(icao, perfData);
        }
        return aircrafts;
    }

    public static HashMap<String, ILS> loadILS(Airport airport) {
        HashMap<String, ILS> approaches = new HashMap<>();
        FileHandle handle = Gdx.files.internal("game/" + TerminalControl.radarScreen.mainName + "/" + TerminalControl.radarScreen.airac + "/ils" + airport.getIcao() + ".ils");
        String[] indivApches = handle.readString().split("\\r?\\n");
        for (String s: indivApches) {
            //For each approach
            if (s.contains("ILS")) {
                ILS ils = new ILS(airport, s);
                approaches.put(ils.getRwy().getName(), ils);
            } else if (s.contains("LDA")) {
                LDA lda = new LDA(airport, s);
                approaches.put(lda.getRwy().getName(), lda);
            } else {
                Gdx.app.log("ILS error", "Invalid approach type specified for " + s);
            }
        }
        return approaches;
    }

    public static HashMap<String, HoldingPoints> loadHoldingPoints(Airport airport) {
        HashMap<String, HoldingPoints> holdingPoints = new HashMap<>();
        FileHandle handle = Gdx.files.internal("game/" + TerminalControl.radarScreen.mainName + "/" + TerminalControl.radarScreen.airac + "/hold" + airport.getIcao() + ".hold");
        JSONObject jo = new JSONObject(handle.readString());
        for (String point: jo.keySet()) {
            holdingPoints.put(point, new HoldingPoints(point, jo.getJSONObject(point)));
        }

        return holdingPoints;
    }

    public static HashMap<String, MissedApproach> loadMissedInfo(Airport airport) {
        HashMap<String, MissedApproach> missedApproaches = new HashMap<>();
        FileHandle handle = Gdx.files.internal("game/" + TerminalControl.radarScreen.mainName + "/" + TerminalControl.radarScreen.airac + "/missedApch" + airport.getIcao() + ".miss");
        JSONObject jo = new JSONObject(handle.readString());
        for (String missed: jo.keySet()) {
            missedApproaches.put(missed, new MissedApproach(missed, airport, jo.getJSONObject(missed)));
        }
        return missedApproaches;
    }

    public static JSONArray loadSaves() {
        JSONArray saves = new JSONArray();

        FileHandle handle = getExtDir("saves/saves.saves");

        if (handle != null && handle.exists()) {
            Array<String> saveIds = new Array<>(handle.readString().split(","));
            if ("".equals(saveIds.get(0))) {
                //"" returned if saves.saves is empty
                return saves;
            }
            for (String id: saveIds) {
                FileHandle handle1 = handle.sibling(id + ".json");
                if (!handle1.exists()) {
                    GameSaver.deleteSave(Integer.parseInt(id));
                    continue;
                }
                String saveString = handle1.readString();
                if (saveString.length() == 0) {
                    GameSaver.deleteSave(Integer.parseInt(id));
                    continue;
                }
                try {
                    saveString = Base64Coder.decodeString(saveString);
                    JSONObject save = new JSONObject(saveString);
                    saves.put(save);
                } catch (JSONException e) {
                    e.printStackTrace();
                    ErrorHandler.sendSaveErrorNoThrow(e, saveString);
                    TerminalControl.toastManager.jsonParseFail();
                    Gdx.app.log("Corrupted save", "JSON parse failure");
                    if (GameSaver.deleteSave(Integer.parseInt(id))) Gdx.app.log("Save deleted", "Corrupted save deleted");
                } catch (IllegalArgumentException e) {
                    e.printStackTrace();
                    ErrorHandler.sendSaveErrorNoThrow(e, saveString);
                    TerminalControl.toastManager.jsonParseFail();
                    Gdx.app.log("Corrupted save", "Base64 decode failure");
                    if (GameSaver.deleteSave(Integer.parseInt(id))) Gdx.app.log("Save deleted", "Corrupted save deleted");
                }
            }
        }

        return saves;
    }

    public static boolean checkIfSaveExists() {
        FileHandle handle = getExtDir("saves/saves.saves");
        if (handle != null && handle.exists()) {
            Array<String> saveIds = new Array<>(handle.readString().split(","));
            //"" returned if saves.saves is empty
            return !"".equals(saveIds.get(0));
        }
        return false;
    }

    public static HashMap<Integer, String> loadAirlines(String icao) {
        HashMap<Integer, String> airlines = new HashMap<>();
        String[] info = Gdx.files.internal("game/" + TerminalControl.radarScreen.mainName + "/" + TerminalControl.radarScreen.airac + "/airlines" + icao + ".airl").readString().split("\\r?\\n");
        int counter = 0;
        for (String indivAirline: info) {
            String callsign = indivAirline.split(",")[0];
            int times = Integer.parseInt(indivAirline.split(",")[1]);
            for (int i = 0; i < times; i++) {
                airlines.put(counter, callsign);
                counter++;
            }
        }

        return airlines;
    }

    public static HashMap<String, String> loadAirlineAircrafts(String icao) {
        HashMap<String, String> aircrafts = new HashMap<>();
        String[] info = Gdx.files.internal("game/" + TerminalControl.radarScreen.mainName + "/" + TerminalControl.radarScreen.airac + "/airlines" + icao + ".airl").readString().split("\\r?\\n");
        for (String indivAirline: info) {
            aircrafts.put(indivAirline.split(",")[0], indivAirline.split(",")[2]);
        }

        return aircrafts;
    }

    public static HashMap<String, String> loadIcaoCallsigns() {
        HashMap<String, String> callsigns = new HashMap<>();
        String[] info = Gdx.files.internal("game/aircrafts/callsigns.call").readString().split("\\r?\\n");
        for (String indivAirline: info) {
            callsigns.put(indivAirline.split(",")[0], indivAirline.split(",")[1]);
        }
        return callsigns;
    }

    public static HashMap<String, Boolean> loadNoise(String icao, boolean sid) {
        HashMap<String, Boolean> noise = new HashMap<>();
        String fileName = sid ? "/noiseSid" : "/noiseStar";
        JSONObject jo = new JSONObject(Gdx.files.internal("game/" + TerminalControl.radarScreen.mainName + "/" + TerminalControl.radarScreen.airac + fileName + icao + ".noi").readString());
        JSONArray nightArray = jo.getJSONArray("night");
        for (int i = 0; i < nightArray.length(); i++) {
            noise.put(nightArray.getString(i), true);
        }
        JSONArray dayArray = jo.getJSONArray("day");
        for (int i = 0; i < dayArray.length(); i++) {
            noise.put(dayArray.getString(i), false);
        }

        return noise;
    }

    public static Array<Array<Integer>> loadShoreline() {
        Array<Array<Integer>> shoreline = new Array<>();
        String[] info = Gdx.files.internal("game/" + TerminalControl.radarScreen.mainName + "/" + TerminalControl.radarScreen.airac + "/shoreline.shore").readString().split("\\r?\\n");
        if ("".equals(info[0])) return shoreline;

        Array<Integer> landmass = new Array<>();
        for (String line: info) {
            if (line.charAt(0) == '\"') {
                if (landmass.size > 0) {
                    shoreline.add(landmass);
                    landmass = new Array<>();
                }
            } else {
                String[] data = line.split(" ");
                landmass.add(Integer.parseInt(data[0])); //x coordinate
                landmass.add(Integer.parseInt(data[1])); //y coordinate
            }
        }

        if (landmass.size > 0) shoreline.add(landmass); //Add the last landmass

        return shoreline;
    }

    public static JSONObject loadSettings() {
        JSONObject settings = null;
        try {
            FileHandle handle = getExtDir("settings.json");
            if (handle != null && handle.exists()) {
                settings = new JSONObject(handle.readString());
            }
        } catch (JSONException e) {
            Gdx.app.log("Corrupted settings", "JSON parse failure");
        }

        return settings;
    }

    public static FileHandle getExtDir(String path) {
        FileHandle handle = null;
        if (Gdx.app.getType() == Application.ApplicationType.Desktop) {
            //If desktop, save to external roaming appData
            if (!TerminalControl.full) {
                handle = Gdx.files.external("AppData/Roaming/TerminalControl/" + path);
            } else {
                handle = Gdx.files.external("AppData/Roaming/TerminalControlFull/" + path);
            }
        } else if (Gdx.app.getType() == Application.ApplicationType.Android) {
            //If Android, check first if local storage available
            if (Gdx.files.isLocalStorageAvailable()) {
                handle = Gdx.files.local(path);
            } else {
                Gdx.app.log("Storage error", "Local storage unavailable for Android!");
                TerminalControl.toastManager.readStorageFail();
            }
        } else {
            Gdx.app.log("Storage error", "Unknown platform!");
        }
        return handle;
    }
}

File: core/src/com/bombbird/terminalcontrol/utilities/files/FileLoader.kt
package com.bombbird.terminalcontrol.utilities.files

import com.badlogic.gdx.Application
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.files.FileHandle
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.Base64Coder
import com.bombbird.terminalcontrol.TerminalControl
import com.bombbird.terminalcontrol.entities.airports.Airport
import com.bombbird.terminalcontrol.entities.approaches.*
import com.bombbird.terminalcontrol.entities.obstacles.CircleObstacle
import com.bombbird.terminalcontrol.entities.obstacles.Obstacle
import com.bombbird.terminalcontrol.entities.obstacles.PolygonObstacle
import com.bombbird.terminalcontrol.entities.procedures.MissedApproach
import com.bombbird.terminalcontrol.entities.procedures.holding.HoldingPoints
import com.bombbird.terminalcontrol.entities.runways.Runway
import com.bombbird.terminalcontrol.entities.sidstar.Sid
import com.bombbird.terminalcontrol.entities.sidstar.Star
import com.bombbird.terminalcontrol.entities.waypoints.Waypoint
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.*

object FileLoader {
    lateinit var mainDir: String
    
    fun loadObstacles(): Array<Obstacle> {
        val obsArray = Array<Obstacle>()
        TerminalControl.radarScreen?.let {
            val obstacles = Gdx.files.internal("game/" + it.mainName + "/" + it.airac + "/obstacle.obs")
            val indivObs = obstacles.readString().split("\\r?\\n".toRegex()).toTypedArray()
            for (s in indivObs) {
                if ("" == s) break
                if (s[0] == '#') continue
                //For each individual obstacle:
                val obs = PolygonObstacle(s)
                obsArray.add(obs)
                it.stage.addActor(obs)
            }
            val restrictions = Gdx.files.internal("game/" + it.mainName + "/" + it.airac + "/restricted.restr")
            val indivRests = restrictions.readString().split("\\r?\\n".toRegex()).toTypedArray()
            for (s in indivRests) {
                if ("" == s) break
                if (s[0] == '#') continue
                //For each individual restricted area
                val area = CircleObstacle(s)
                obsArray.add(area)
                it.stage.addActor(area)
            }
        }
        return obsArray
    }

    fun loadWaypoints(): HashMap<String, Waypoint> {
        val waypoints = HashMap<String, Waypoint>()
        TerminalControl.radarScreen?.let {
            val handle = Gdx.files.internal("game/" + it.mainName + "/" + it.airac + "/waypoint.way")
            val wayptStr = handle.readString()
            val indivWpt = wayptStr.split("\\r?\\n".toRegex()).toTypedArray()
            for (s in indivWpt) {
                //For each waypoint
                var name = ""
                var x = 0
                var y = 0
                for ((index, s1) in s.split(" ".toRegex()).toTypedArray().withIndex()) {
                    when (index) {
                        0 -> name = s1
                        1 -> x = s1.toInt()
                        2 -> y = s1.toInt()
                        else -> Gdx.app.log("Load error", "Unexpected additional parameter in game/" + it.mainName + "/" + it.airac + "/waypoint.way -> $name")
                    }
                }
                val waypoint = Waypoint(name, x, y)
                waypoints[name] = waypoint
                it.stage.addActor(waypoint)
            }
        }
        return waypoints
    }

    fun loadRunways(icao: String): HashMap<String, Runway> {
        val runways = HashMap<String, Runway>()
        TerminalControl.radarScreen?.let {
            val handle = Gdx.files.internal("game/" + it.mainName + "/" + it.airac + "/runway" + icao + ".rwy")
            val indivRwys = handle.readString().split("\\r?\\n".toRegex()).toTypedArray()
            for (s in indivRwys) {
                //For each individual runway
                val runway = Runway(s)
                runways[runway.name] = runway
            }
        }
        return runways
    }

    fun loadStars(airport: Airport): HashMap<String, Star> {
        //Load STARs
        val stars = HashMap<String, Star>()
        TerminalControl.radarScreen?.let {
            val handle = Gdx.files.internal("game/" + it.mainName + "/" + it.airac + "/star" + airport.icao + ".star")
            val jo = JSONObject(handle.readString())
            for (name in jo.keySet()) {
                //For each individual STAR
                val star = Star(airport, jo.getJSONObject(name))
                star.name = name
                stars[name] = star
            }
        }
        return stars
    }

    fun loadSids(airport: Airport): HashMap<String, Sid> {
        //Load SIDs
        val sids = HashMap<String, Sid>()
        TerminalControl.radarScreen?.let {
            val handle = Gdx.files.internal("game/" + it.mainName + "/" + it.airac + "/sid" + airport.icao + ".sid")
            val jo = JSONObject(handle.readString())
            for (name in jo.keySet()) {
                //For each individual SID
                val sid = Sid(airport, jo.getJSONObject(name))
                sid.name = name
                sids[name] = sid
            }
        }
        return sids
    }

    fun loadAircraftData(): HashMap<String, IntArray> {
        val aircrafts = HashMap<String, IntArray>()
        val handle = Gdx.files.internal("game/aircrafts/aircrafts.air")
        val indivAircrafts = handle.readString().split("\\r?\\n".toRegex()).toTypedArray()
        for (s in indivAircrafts) {
            if (s[0] == '#') {
                continue
            }
            //For each aircraft
            var index = 0
            var icao = ""
            val perfData = IntArray(8)
            for (s1 in s.split(",".toRegex()).toTypedArray()) {
                if (index == 0) {
                    icao = s1 //Icao code of aircraft
                } else if (index in 2..5 || index in 7..8) {
                    perfData[index - 1] = s1.toInt()
                } else if (index == 1 || index == 6) {
                    //Wake turb cat, recat
                    val cat = s1[0]
                    perfData[index - 1] = cat.code
                    if (index == 1 && !(cat == 'M' || cat == 'H' || cat == 'J')) Gdx.app.log("Wake cat error", "Unknown category for $icao")
                    if (index == 6 && cat !in 'A'..'F') Gdx.app.log("Recat error", "Unknown category for $icao")
                } else {
                    Gdx.app.log("Load error", "Unexpected additional parameter in game/aircrafts.air")
                }
                index++
            }
            if (index <= perfData.size) Gdx.app.log("Load error", "Possible missing aircraft parameter for $icao")
            aircrafts[icao] = perfData
        }
        return aircrafts
    }

    fun loadApch(airport: Airport): HashMap<String, Approach> {
        val approaches = HashMap<String, Approach>()
        TerminalControl.radarScreen?.let {
            val handle = Gdx.files.internal("game/" + it.mainName + "/" + it.airac + "/apch" + airport.icao + ".apch")
            val jo = JSONObject(handle.readString())
            for (name in jo.keySet()) {
                //For each approach
                val data = jo.getJSONObject(name)
                when(val type = data.getString("type")) {
                    "ILS" -> {
                        val ils = ILS(airport, name, data)
                        approaches[ils.rwy.name] = ils
                    }
                    "LDA", "IGS" -> {
                        val lda = OffsetILS(airport, name, data)
                        approaches[lda.rwy.name] = lda
                    }
                    "CIRCLING" -> {
                        val circle = Circling(airport, name, data)
                        approaches[circle.rwy.name] = circle
                    }
                    "RNP" -> {
                        val rnp = RNP(airport, name, data)
                        approaches[rnp.rwy.name] = rnp
                    }
                    else -> Gdx.app.log("ILS error", "Invalid approach type $type specified")
                }
            }
        }
        return approaches
    }

    fun loadHoldingPoints(airport: Airport): HashMap<String, HoldingPoints> {
        val holdingPoints = HashMap<String, HoldingPoints>()
        TerminalControl.radarScreen?.let {
            val handle = Gdx.files.internal("game/" + it.mainName + "/" + it.airac + "/hold" + airport.icao + ".hold")
            val jo = JSONObject(handle.readString())
            for (point in jo.keySet()) {
                holdingPoints[point] = HoldingPoints(point, jo.getJSONObject(point))
            }
        }
        return holdingPoints
    }

    fun loadMissedInfo(airport: Airport): HashMap<String, MissedApproach> {
        val missedApproaches = HashMap<String, MissedApproach>()
        TerminalControl.radarScreen?.let {
            val handle = Gdx.files.internal("game/" + it.mainName + "/" + it.airac + "/missedApch" + airport.icao + ".miss")
            val jo = JSONObject(handle.readString())
            for (missed in jo.keySet()) {
                missedApproaches[missed] = MissedApproach(jo.getJSONObject(missed))
            }
        }
        return missedApproaches
    }

    fun loadSaves(): JSONArray {
        val saves = JSONArray()
        val handle = getExtDir("saves/saves.saves")
        if (handle != null && handle.exists()) {
            val saveIds = Array(handle.readString().split(",".toRegex()).toTypedArray())
            if ("" == saveIds[0]) {
                //"" returned if saves.saves is empty
                return saves
            }
            for (id in saveIds) {
                val handle1 = handle.sibling("$id.json")
                if (!handle1.exists()) {
                    GameSaver.deleteSave(id.toInt())
                    continue
                }
                var saveString = handle1.readString()
                if (saveString.isEmpty()) {
                    GameSaver.deleteSave(id.toInt())
                    continue
                }
                try {
                    saveString = Base64Coder.decodeString(saveString)
                    val save = JSONObject(saveString)
                    saves.put(save)
                } catch (e: JSONException) {
                    e.printStackTrace()
                    //ErrorHandler.sendSaveErrorNoThrow(e, saveString);
                    TerminalControl.toastManager.jsonParseFail()
                    Gdx.app.log("Corrupted save", "JSON parse failure")
                    if (GameSaver.deleteSave(id.toInt())) Gdx.app.log("Save deleted", "Corrupted save deleted")
                } catch (e: IllegalArgumentException) {
                    e.printStackTrace()
                    //ErrorHandler.sendSaveErrorNoThrow(e, saveString);
                    TerminalControl.toastManager.jsonParseFail()
                    Gdx.app.log("Corrupted save", "Base64 decode failure")
                    if (GameSaver.deleteSave(id.toInt())) Gdx.app.log("Save deleted", "Corrupted save deleted")
                }
            }
        }
        return saves
    }

    fun checkIfSaveExists(): Boolean {
        val handle = getExtDir("saves/saves.saves")
        return if (handle != null && handle.exists()) {
            //"" returned if saves.saves is empty
            "" != handle.readString()
        } else false
    }

    fun loadAirlines(icao: String): HashMap<Int, String> {
        val airlines = HashMap<Int, String>()
        TerminalControl.radarScreen?.let {
            val info = Gdx.files.internal("game/" + it.mainName + "/" + it.airac + "/airlines" + icao + ".airl").readString().split("\\r?\\n".toRegex()).toTypedArray()
            var counter = 0
            for (indivAirline in info) {
                val callsign = indivAirline.split(",".toRegex()).toTypedArray()[0]
                val times = indivAirline.split(",".toRegex()).toTypedArray()[1].toInt()
                for (i in 0 until times) {
                    airlines[counter] = callsign
                    counter++
                }
            }
        }
        return airlines
    }

    fun loadAirlineAircrafts(icao: String): HashMap<String, String> {
        val aircrafts = HashMap<String, String>()
        TerminalControl.radarScreen?.let {
            val info = Gdx.files.internal("game/" + it.mainName + "/" + it.airac + "/airlines" + icao + ".airl").readString().split("\\r?\\n".toRegex()).toTypedArray()
            for (indivAirline in info) {
                aircrafts[indivAirline.split(",".toRegex()).toTypedArray()[0]] = indivAirline.split(",".toRegex()).toTypedArray()[2]
            }
        }
        return aircrafts
    }

    fun loadIcaoCallsigns(): HashMap<String, String> {
        val callsigns = HashMap<String, String>()
        val info = Gdx.files.internal("game/aircrafts/callsigns.call").readString().split("\\r?\\n".toRegex()).toTypedArray()
        for (indivAirline in info) {
            callsigns[indivAirline.split(",".toRegex()).toTypedArray()[0]] = indivAirline.split(",".toRegex()).toTypedArray()[1]
        }
        return callsigns
    }

    fun loadNoise(icao: String, sid: Boolean): HashMap<String, Boolean> {
        val noise = HashMap<String, Boolean>()
        TerminalControl.radarScreen?.let {
            val fileName = if (sid) "/noiseSid" else "/noiseStar"
            val jo = JSONObject(Gdx.files.internal("game/" + it.mainName + "/" + it.airac + fileName + icao + ".noi").readString())
            val nightArray = jo.getJSONArray("night")
            for (i in 0 until nightArray.length()) {
                noise[nightArray.getString(i)] = true
            }
            val dayArray = jo.getJSONArray("day")
            for (i in 0 until dayArray.length()) {
                noise[dayArray.getString(i)] = false
            }
        }
        return noise
    }

    fun loadShoreline(): Array<Array<Int>> {
        val shoreline = Array<Array<Int>>()
        TerminalControl.radarScreen?.let {
            val info = Gdx.files.internal("game/" + it.mainName + "/" + it.airac + "/shoreline.shore").readString().split("\\r?\\n".toRegex()).toTypedArray()
            if ("" == info[0]) return shoreline
            var landmass = Array<Int>()
            for (line in info) {
                if (line[0] == '\"') {
                    if (landmass.size > 0) {
                        shoreline.add(landmass)
                        landmass = Array()
                    }
                } else {
                    val data = line.split(" ".toRegex()).toTypedArray()
                    landmass.add(data[0].toInt()) //x coordinate
                    landmass.add(data[1].toInt()) //y coordinate
                }
            }
            if (landmass.size > 0) shoreline.add(landmass) //Add the last landmass
        }
        return shoreline
    }

    fun loadSettings(): JSONObject? {
        var settings: JSONObject? = null
        try {
            val handle = getExtDir("settings.json")
            if (handle != null && handle.exists()) {
                settings = JSONObject(handle.readString())
            }
        } catch (e: JSONException) {
            Gdx.app.log("Corrupted settings", "JSON parse failure")
        }
        return settings
    }

    fun loadStats(): JSONObject? {
        var stats: JSONObject? = null
        try {
            val handle = getExtDir("stats.json")
            if (handle != null && handle.exists()) {
                stats = JSONObject(handle.readString())
            }
        } catch (e: JSONException) {
            Gdx.app.log("Corrupted settings", "JSON parse failure")
        }
        return stats
    }

    fun getAvailableDatatagConfigs(): kotlin.Array<String> {
        val handle = getExtDir("datatags")
        return if (handle != null && handle.exists() && handle.isDirectory) {
            handle.list().map { it.name() }.toTypedArray()
        } else {
            handle?.mkdirs()
            arrayOf()
        }
    }

    fun getDatatagLayout(name: String): JSONObject? {
        val handle = getExtDir("datatags/$name")
        return if (handle != null && handle.exists()) JSONObject(handle.readString()) else null
    }

    fun getExtDir(path: String): FileHandle? {
        var handle: FileHandle? = null
        if (Gdx.app.type == Application.ApplicationType.Desktop) {
            //If desktop, save to external roaming appData
            handle = if (!TerminalControl.full) {
                Gdx.files.external("AppData/Roaming/TerminalControl/$path")
            } else {
                Gdx.files.external("AppData/Roaming/TerminalControlFull/$path")
            }
        } else if (Gdx.app.type == Application.ApplicationType.Android) {
            //If Android, check first if local storage available
            if (Gdx.files.isLocalStorageAvailable) {
                handle = Gdx.files.local(path)
            } else {
                Gdx.app.log("Storage error", "Local storage unavailable for Android!")
                TerminalControl.toastManager.readStorageFail()
            }
        } else {
            Gdx.app.log("Storage error", "Unknown platform!")
        }
        return handle
    }
}